File: frontend/src/api/services/boardService.ts
import api from '../axios';

export const boardService = {
  getBoards: async () => {
    const response = await api.get('/boards');
    return response.data;
  },

  createBoard: async (boardData: { title: string; category?: string; description?: string }) => {
    const response = await api.post('/boards', boardData);
    return response.data;
  },

  getBoardById: async (id: string) => {
    const response = await api.get(`/boards/${id}`);
    return response.data;
  },

  joinBoardByCode: async (code: string) => {
    const response = await api.get(`/boards/join/${code}`);
    return response.data;
  },

  setJoinCode: async (id: string, code: string) => {
    const response = await api.post(`/boards/${id}/joinCode`, { joinCode: code });
    return response.data;
  },

  updateBoard: async (id: string, boardData: any) => {
    const response = await api.put(`/boards/${id}`, boardData);
    return response.data;
  },

  deleteBoard: async (id: string) => {
    const response = await api.delete(`/boards/${id}`);
    return response.data;
  },

  saveYjsState: async (id: string, update: Uint8Array) => {
    const response = await api.put(`/boards/${id}/yjs`, update.buffer.slice(update.byteOffset, update.byteOffset + update.byteLength), {
      headers: { 'Content-Type': 'application/octet-stream' },
      transformRequest: (data) => data, // bypass axios JSON serialization
    });
    return response.data;
  },

  // keepalive: true ensures the request completes even during page reload/close
  saveYjsStateOnUnload: (id: string, update: Uint8Array) => {
    const buffer = update.buffer.slice(update.byteOffset, update.byteOffset + update.byteLength) as ArrayBuffer;
    const baseURL = import.meta.env.VITE_API_BASE_URL || '';
    fetch(`${baseURL}/boards/${id}/yjs`, {
      method: 'PUT',
      body: buffer,
      headers: { 'Content-Type': 'application/octet-stream' },
      keepalive: true,
      credentials: 'include',
    });
  },
};

File: flutterink/lib/api/board_service.dart
import 'package:dio/dio.dart';
import './api.dart';
import '../models/board.dart';

class BoardService {
  final Api _api = Api.instance;

  // GET /api/boards
  Future<List<Board>> getBoards() async {
    try {
      final response = await _api.get('/boards');
      return (response.data as List)
          .map((json) => Board.fromJson(json as Map<String, dynamic>))
          .toList();
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to fetch boards';
      throw Exception(message);
    }
  }

  // POST /api/boards
  Future<Board> createBoard({
    required String title,
    String? category,
    String? description,
  }) async {
    try {
      final response = await _api.post('/boards', data: {
        'title': title,
        if (category != null) 'category': category,
        if (description != null) 'description': description,
      });
      return Board.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to create board';
      throw Exception(message);
    }
  }

  // GET /api/boards/:id
  Future<Board> getBoardById(String id) async {
    try {
      final response = await _api.get('/boards/$id');
      return Board.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to fetch board';
      throw Exception(message);
    }
  }

  // GET /api/boards/join/:code
  Future<Board> joinBoardByCode(String code) async {
    try {
      final response = await _api.get('/boards/join/$code');
      return Board.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to join board';
      throw Exception(message);
    }
  }

  // POST /api/boards/:id/joinCode
  Future<Board> setJoinCode(String id, String code) async {
    try {
      final response = await _api.post('/boards/$id/joinCode', data: {
        'joinCode': code,
      });
      return Board.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to set join code';
      throw Exception(message);
    }
  }

  // PUT /api/boards/:id
  Future<Board> updateBoard(String id, Map<String, dynamic> boardData) async {
    try {
      final response = await _api.put('/boards/$id', data: boardData);
      return Board.fromJson(response.data as Map<String, dynamic>);
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to update board';
      throw Exception(message);
    }
  }

  // DELETE /api/boards/:id
  Future<void> deleteBoard(String id) async {
    try {
      await _api.delete('/boards/$id');
    } on DioException catch (e) {
      final message = e.response?.data['error'] ?? 'Failed to delete board';
      throw Exception(message);
    }
  }
}
